adapters/sn2core/sn2core.go

```diff
@@ -165,24 +169,39 @@ func AdaptDeclareTransaction(t *starknet.Transaction) *core.DeclareTransaction {
 		Version:               (*core.TransactionVersion)(t.Version),
 		ClassHash:             t.ClassHash,
 		CompiledClassHash:     t.CompiledClassHash,
-		ResourceBounds:        adaptResourceBounds(t.ResourceBounds),
-		Tip:                   t.Tip,
+		ResourceBounds:        resourceBounds,
+		Tip:                   safeFeltToUint64(t.Tip),
 		PaymasterData:         t.PaymasterData,
 		AccountDeploymentData: t.AccountDeploymentData,
 		NonceDAMode:           core.DataAvailabilityMode(t.NonceDAMode),
 		FeeDAMode:             core.DataAvailabilityMode(t.FeeDAMode),
-	}
+	}, nil
 }
 
-func adaptResourceBounds(rb map[starknet.Resource]*starknet.ResourceBounds) map[core.Resource]*core.ResourceBounds {
+func adaptResourceBounds(rb map[starknet.Resource]*starknet.ResourceBounds) (map[core.Resource]*core.ResourceBounds, error) {
 	coreBounds := make(map[core.Resource]*core.ResourceBounds, len(rb))
 	for resource, bounds := range rb {
-		coreBounds[core.Resource(resource)] = &core.ResourceBounds{
-			MaxAmount:       bounds.MaxAmount,
+		coreResource, err := adaptResource(resource)
+		if err != nil {
+			return nil, err
+		}
+		coreBounds[coreResource] = &core.ResourceBounds{
+			MaxAmount:       bounds.MaxAmount.Uint64(),
 			MaxPricePerUnit: bounds.MaxPricePerUnit,
 		}
 	}
-	return coreBounds
+	return coreBounds, nil
+}
+
+func adaptResource(resource starknet.Resource) (core.Resource, error) {
+	switch resource {
+	case starknet.ResourceL1Gas:
+		return core.ResourceL1Gas, nil
+	case starknet.ResourceL2Gas:
+		return core.ResourceL2Gas, nil
+	default:
+		return 0, fmt.Errorf("unknown starknet.Resource %s", resource)
```

Comment:
I think it makes sense to keep the UnmarshalJSON function to catch this. So we don't need to make `AdaptInvokeTransaction` return an error and bubble it up to the caller.

Reply:
`AdaptInvokeTransaction` is public, so we could encounter an invalid `starknet.Resource` even if an `UnmarshalJSON` existed.

`AdaptTransaction` already returns an error so it doesn't matter much as far as Juno goes.

Reply:
Never mind, I see what you mean. For some reason unmarshalling works when I include `UnmarshalText`. The `encoding/json` docs are vague/misleading about this:

> The map's key type must either be any string type, an integer, implement json.Unmarshaler, or implement encoding.TextUnmarshaler.

Anyways, should be fixed now.